[5.15.18] Fix detection of OpenGL Qt backend and OpenSSL linkage on macOS

jcfr opened the pull request:
- Ensures proper detection of the "Desktop OpenGL" backend by removing unnecessary AGL framework linkage.
- Add `-headerpad_max_install_names` to OpenSSL linkage: Addresses errors related to updating install names or rpaths due to insufficient header padding.
- Speed up source downloads of qtlocation and qtwebengine repositories by using shallow clones with minimal history and filtered blobs.
- Updating the `qtwebengine` version to `v5.15.19-lts`.
- Incorporation of multiple patches adapted from macports-ports.
- Addition of verbose flag to git apply for better feedback.

cc: @jamesobutler @ebrahimebrahim

jcfr committed:
patches: Remove AGL framework linkage in macOS configuration

Ensure detection of "Desktop OpenGL" backend works as expected by
removing the AGL framework linkage.

This change improves the configuration as verified by the following command:

```
./qt-easy-build/Build-qt.sh -j 8 -s macosx26.1 -d 13
```

and the resulting configuration output:

```
  OpenGL:
    Desktop OpenGL ....................... yes
    OpenGL ES 2.0 ........................ no
    OpenGL ES 3.0 ........................ no
    OpenGL ES 3.1 ........................ no
    OpenGL ES 3.2 ........................ no
```

Based of the following:
* https://trac.macports.org/ticket/72729
* https://github.com/qt/qtbase/commit/cdb33c3d5621ce035ad6950c8e2268fe94b73de5
* https://github.com/macports/macports-ports/pull/29396/commits/15c9bafaffe4a49c8f08837c83db0683c87a8286

jcfr committed:
Build-qt.sh: Add -headerpad_max_install_names to OpenSSL linkage

This addresses errors like the following where updating install names
or rpaths fails  due to insufficient header padding:

```
+ install_name_tool -id /path/to/qt-everywhere-deps-5.15.18/openssl-1.1.1w/libcrypto.dylib /path/to/qt-everywhere-deps-5.15.18/openssl-1.1.1w/libcrypto.dylib
+ install_name_tool -change /usr/local/ssl/lib/libcrypto.1.0.0.dylib /path/to/qt-everywhere-deps-5.15.18/openssl-1.1.1w/libcrypto.dylib -id /path/to/qt-everywhere-deps-5.15.18/openssl-1.1.1w/libssl.dylib /path/to/qt-everywhere-deps-5.15.18/openssl-1.1.1w/libssl.dylib
error: /Applications/Xcode-26.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: changing install names or rpaths can't be redone for: /path/to/qt-everywhere-deps-5.15.18/openssl-1.1.1w/libssl.dylib (for architecture x86_64) because larger updated load commands do not fit (the program must be relinked, and you may need to use -headerpad or -headerpad_max_install_names)
```

jcfr committed:
Build-qt.sh: Speed up download of qtlocation and qtwebengine sources

This change improves the download speed of the qtlocation and qtwebengine
source repositories by using shallow clones with minimal history and
filtered blobs.

jamesobutler commented:
The relevant qtbase commit is https://github.com/qt/qtbase/commit/cdb33c3d5621ce035ad6950c8e2268fe94b73de5 that came out of https://bugreports.qt.io/browse/QTBUG-137687. Since this commit came after Qt 5.15 went EOL, they’ve stopped back porting these type of fixes to the Qt 5.15 branches.

jcfr committed:
Build-qt.sh: Download qtwebengine v5.15.19-lts instead of v5.15.18-lts-lgpl

jcfr committed:
Build-qt.sh: Add patches based of macports-ports and update logic to apply

Adds multiple patches adapted from macports-ports and updates Build-qt.sh
script to apply these patches during the build process.

Patches have been updated to reference expected top-level directory (e.g. qtbase)

jcfr committed:
Build-qt.sh: Exclude obsolete patches

jcfr committed:
fix: Update patches for git compatibility

This updates various patch files to use "a/" and "b/" prefixes and
removes ".orig" suffixes, enabling them to be applied using git.

jcfr committed:
Build-qt.sh: Add verbose flag to git apply

Provides additional feedback during the patch application process.

jcfr commented:
> they’ve stopped back porting these type of fixes to the Qt 5.15 branches

Ditto. We should now be good ... with this set of patches, the build is now moving forward :rocket:

jcfr committed:
Build-qt.sh: Add missing patch to support building boost bundled in qtlocation

Patch was updated to include the prefix "qtlocation/src/3rdparty/mapbox-gl-native/deps/boost/1.65.1/include/"

Copied from https://github.com/macports/macports-ports/blob/3f0edb5af9ea0763975db46f78cb0093b9445c2f/devel/boost171/files/patch-boost-clang16-cpp17-compat.diff

jcfr committed:
Build-qt.sh: Re-enable "0003-qtwebengine-clang16.patch" and update to support 5.15.19

jcfr committed:
Build-qt.sh: Fix libpng build error associated with qtwebengine

Adapted from https://github.com/macports/macports-ports/blob/3f0edb5af9ea0763975db46f78cb0093b9445c2f/aqua/qt5/files/patch-qtwebengine-libpng.diff

Fix errors like the following:

```
../../3rdparty/chromium/third_party/libpng/pngpriv.h:527:16: fatal error: 'fp.h' file not found
  527 | #      include <fp.h>
      |                ^~~~~~
1 error generated.
```

jcfr commented:
:crossed_fingers:

jcfr commented:
Since I forgot to apply yet another patch to rollback the version of the qtwebengine module ... I will manually apply the changes on `computron`:

### Before

```
$ pwd
/D/Support/qt-everywhere-build-5.15.18/lib/cmake

$ fgrep -risnI "5.15.19" . 
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:8:set(Qt5WebEngineCore_VERSION_STRING 5.15.19)
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:58:        "${_qt5WebEngineCore_install_prefix}/lib/QtWebEngineCore.framework/Versions/5/Headers/5.15.19/"
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:59:        "${_qt5WebEngineCore_install_prefix}/lib/QtWebEngineCore.framework/Versions/5/Headers/5.15.19/QtWebEngineCore"
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:94:                5.15.19 ${_Qt5WebEngineCore_FIND_VERSION_EXACT}
./Qt5WebEngineCore/Qt5WebEngineCoreConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.19)
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:8:set(Qt5PdfWidgets_VERSION_STRING 5.15.19)
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:58:        "${_qt5PdfWidgets_install_prefix}/lib/QtPdfWidgets.framework/Versions/5/Headers/5.15.19/"
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:59:        "${_qt5PdfWidgets_install_prefix}/lib/QtPdfWidgets.framework/Versions/5/Headers/5.15.19/QtPdfWidgets"
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:94:                5.15.19 ${_Qt5PdfWidgets_FIND_VERSION_EXACT}
./Qt5PdfWidgets/Qt5PdfWidgetsConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.19)
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.19)
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:8:set(Qt5WebEngineWidgets_VERSION_STRING 5.15.19)
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:58:        "${_qt5WebEngineWidgets_install_prefix}/lib/QtWebEngineWidgets.framework/Versions/5/Headers/5.15.19/"
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:59:        "${_qt5WebEngineWidgets_install_prefix}/lib/QtWebEngineWidgets.framework/Versions/5/Headers/5.15.19/QtWebEngineWidgets"
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:94:                5.15.19 ${_Qt5WebEngineWidgets_FIND_VERSION_EXACT}
./Qt5Pdf/Qt5PdfConfig.cmake:8:set(Qt5Pdf_VERSION_STRING 5.15.19)
./Qt5Pdf/Qt5PdfConfig.cmake:58:        "${_qt5Pdf_install_prefix}/lib/QtPdf.framework/Versions/5/Headers/5.15.19/"
./Qt5Pdf/Qt5PdfConfig.cmake:59:        "${_qt5Pdf_install_prefix}/lib/QtPdf.framework/Versions/5/Headers/5.15.19/QtPdf"
./Qt5Pdf/Qt5PdfConfig.cmake:94:                5.15.19 ${_Qt5Pdf_FIND_VERSION_EXACT}
./Qt5Pdf/Qt5PdfConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.19)
./Qt5WebEngine/Qt5WebEngineConfig.cmake:8:set(Qt5WebEngine_VERSION_STRING 5.15.19)
./Qt5WebEngine/Qt5WebEngineConfig.cmake:58:        "${_qt5WebEngine_install_prefix}/lib/QtWebEngine.framework/Versions/5/Headers/5.15.19/"
./Qt5WebEngine/Qt5WebEngineConfig.cmake:59:        "${_qt5WebEngine_install_prefix}/lib/QtWebEngine.framework/Versions/5/Headers/5.15.19/QtWebEngine"
./Qt5WebEngine/Qt5WebEngineConfig.cmake:94:                5.15.19 ${_Qt5WebEngine_FIND_VERSION_EXACT}
./Qt5WebEngine/Qt5WebEngineConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.19)

```

**Update command:**

```
$ pwd
/D/Support/qt-everywhere-build-5.15.18/lib/cmake

$ grep -rl "5.15.19" . \                                                                       
  | xargs sed -i '' 's/5\.15\.19/5.15.18/g'
```

### After

There are no files with `-5.15.19`:

```
$ pwd
/D/Support/qt-everywhere-build-5.15.18/lib/cmake

$ fgrep -risnI "5.15.19" . 
```

.. and the relevant modules have been updated:

```
$ fgrep -risnI "5.15.18" . | grep -E "(Qt5WebEngine|Qt5Pdf)"
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:8:set(Qt5WebEngineCore_VERSION_STRING 5.15.18)
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:58:        "${_qt5WebEngineCore_install_prefix}/lib/QtWebEngineCore.framework/Versions/5/Headers/5.15.18/"
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:59:        "${_qt5WebEngineCore_install_prefix}/lib/QtWebEngineCore.framework/Versions/5/Headers/5.15.18/QtWebEngineCore"
./Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:94:                5.15.18 ${_Qt5WebEngineCore_FIND_VERSION_EXACT}
./Qt5WebEngineCore/Qt5WebEngineCoreConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.18)
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:8:set(Qt5PdfWidgets_VERSION_STRING 5.15.18)
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:58:        "${_qt5PdfWidgets_install_prefix}/lib/QtPdfWidgets.framework/Versions/5/Headers/5.15.18/"
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:59:        "${_qt5PdfWidgets_install_prefix}/lib/QtPdfWidgets.framework/Versions/5/Headers/5.15.18/QtPdfWidgets"
./Qt5PdfWidgets/Qt5PdfWidgetsConfig.cmake:94:                5.15.18 ${_Qt5PdfWidgets_FIND_VERSION_EXACT}
./Qt5PdfWidgets/Qt5PdfWidgetsConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.18)
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.18)
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:8:set(Qt5WebEngineWidgets_VERSION_STRING 5.15.18)
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:58:        "${_qt5WebEngineWidgets_install_prefix}/lib/QtWebEngineWidgets.framework/Versions/5/Headers/5.15.18/"
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:59:        "${_qt5WebEngineWidgets_install_prefix}/lib/QtWebEngineWidgets.framework/Versions/5/Headers/5.15.18/QtWebEngineWidgets"
./Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:94:                5.15.18 ${_Qt5WebEngineWidgets_FIND_VERSION_EXACT}
./Qt5Pdf/Qt5PdfConfig.cmake:8:set(Qt5Pdf_VERSION_STRING 5.15.18)
./Qt5Pdf/Qt5PdfConfig.cmake:58:        "${_qt5Pdf_install_prefix}/lib/QtPdf.framework/Versions/5/Headers/5.15.18/"
./Qt5Pdf/Qt5PdfConfig.cmake:59:        "${_qt5Pdf_install_prefix}/lib/QtPdf.framework/Versions/5/Headers/5.15.18/QtPdf"
./Qt5Pdf/Qt5PdfConfig.cmake:94:                5.15.18 ${_Qt5Pdf_FIND_VERSION_EXACT}
./Qt5Pdf/Qt5PdfConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.18)
./Qt5WebEngine/Qt5WebEngineConfig.cmake:8:set(Qt5WebEngine_VERSION_STRING 5.15.18)
./Qt5WebEngine/Qt5WebEngineConfig.cmake:58:        "${_qt5WebEngine_install_prefix}/lib/QtWebEngine.framework/Versions/5/Headers/5.15.18/"
./Qt5WebEngine/Qt5WebEngineConfig.cmake:59:        "${_qt5WebEngine_install_prefix}/lib/QtWebEngine.framework/Versions/5/Headers/5.15.18/QtWebEngine"
./Qt5WebEngine/Qt5WebEngineConfig.cmake:94:                5.15.18 ${_Qt5WebEngine_FIND_VERSION_EXACT}
./Qt5WebEngine/Qt5WebEngineConfigVersion.cmake:2:set(PACKAGE_VERSION 5.15.18)
```

jcfr commented:
And now configuring Slicer works as expected  :tada:  ...

```
[...]
-- Configuring Slicer with Qt 5.15.18 (using modules: Core, Widgets, Network, OpenGL, PrintSupport, UiTools, Xml, XmlPatterns, Svg, Sql, Multimedia, MultimediaWidgets, WebEngine, WebEngineWidgets, WebChannel, Qml, LinguistTools, Test, )
-- Setting QT_PLUGINS_DIR: /D/Support/qt-everywhere-build-5.15.18/plugins
-- Setting QT_BINARY_DIR: /D/Support/qt-everywhere-build-5.15.18/bin
-- Setting QT_LIBRARY_DIR: /D/Support/qt-everywhere-build-5.15.18/lib
-- Setting Slicer_BUILD_KRB5_GSSAPI_STUB to 'OFF'
-- Setting ExternalData_OBJECT_STORES to '/D/P/A/ExternalData/Objects'
[...]
```

<details><summary>... whereas prior to the change, the following was reported:</summary>
<p>




```
[...]
CMake Warning at CMake/SlicerBlockFindQtAndCheckVersion.cmake:22 (find_package):
  Found package configuration file:

    /D/Support/qt-everywhere-build-5.15.18/lib/cmake/Qt5/Qt5Config.cmake

  but it set Qt5_FOUND to FALSE so package "Qt5" is considered to be NOT
  FOUND.  Reason given by package:

  Failed to find Qt5 component "WebEngine" config file at
  "/D/Support/qt-everywhere-build-5.15.18/lib/cmake/Qt5WebEngine/Qt5WebEngineConfig.cmake"


  Failed to find Qt5 component "WebEngineWidgets" config file at
  "/D/Support/qt-everywhere-build-5.15.18/lib/cmake/Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake"




Call Stack (most recent call first):
  CMake/SlicerBlockFindQtAndCheckVersion.cmake:67 (__SlicerBlockFindQtAndCheckVersion_find_qt)
  CMakeLists.txt:717 (include)


CMake Error at CMake/SlicerBlockFindQtAndCheckVersion.cmake:30 (message):
  error: Qt 5.15 was not found on your system.You probably need to set the
  Qt5_DIR variable.
Call Stack (most recent call first):
  CMake/SlicerBlockFindQtAndCheckVersion.cmake:67 (__SlicerBlockFindQtAndCheckVersion_find_qt)
  CMakeLists.txt:717 (include)


-- Configuring incomplete, errors occurred!
[...]
```

</p>
</details>

jcfr commented:
Test build on macOS against the latest CTK was :white_check_mark: :100: 

That said ... I observed the following linker warnings during the build (see below).

---

I am working on ...
* addressing these warnings
* finalizing the CTK + Qt6 update

Later this evening, once I have integrate relevant changes .. I will manually trigger the Preview & will release tomorrow :rocket: 

---

### Linker Warnings

```
ld: warning: building for macOS-13.0, but linking with dylib '/D/P/A/DCMTK-build/lib/libofstd.18.dylib' which was built for newer version 15.0
ld: warning: building for macOS-13.0, but linking with dylib '/D/P/A/DCMTK-build/lib/liboficonv.18.dylib' which was built for newer version 15.0
ld: warning: building for macOS-13.0, but linking with dylib '/D/P/A/CTK-build/CTK-build/bin/libCTKWidgets.0.1.dylib' which was built for newer version 15.0
ld: warning: building for macOS-13.0, but linking with dylib '/D/P/A/CTK-build/CTK-build/bin/libCTKCore.0.1.dylib' which was built for newer version 15.0
ld: warning: building for macOS-13.0, but linking with dylib '/D/P/A/CTK-build/QtTesting-build/libQtTesting.dylib' which was built for newer version 15.0
[100%] Built target qSlicerVolumeRenderingModuleCxxTests

```

jcfr commented:
Before tagging the release, I understand the link warning.

Running the following command allows to extract the `minos` associated with the `LC_BUILD_VERSION` section (if any) of each library:

```
find . \( -name '*.dylib' -o -name '*.so' \) -print0 | while IFS= read -r -d '' lib; do
  echo "$(otool -l "$lib" | awk '/LC_BUILD_VERSION/{f=1} f&&/minos/{print $2; f=0}') $lib";
done > /tmp/slicer-minos-lib.txt
```

Inspecting the results shows that all relevant libraries are built with the
expected minos value 13.0. The few exceptions (with minos 15.7) are limited to
CMake/compile test artifacts (e.g., ITK/SimpleITK CMakeTmp directories) and are
not part of the final linked Slicer binaries.

```
$ grep -v "^13" /tmp/slicer-minos-lib.txt | grep -v "^ " | grep -v "python3.12"
15.7 ./ITK-build/Modules/Core/Common/CMakeTmp/ITK_PRIVATE_DYNAMIC_CAST/libbase.dylib
15.7 ./SimpleITK-build/SimpleITK-build/Wrapping/Python/CMakeTmp/link_flags/osx_dynamic_lookup/MODULE/STATIC/build/counter.so
15.7 ./SimpleITK-build/SimpleITK-build/CMakeTmp/link_flags/osx_dynamic_lookup/MODULE/SHARED/build/counter.so
15.7 ./SimpleITK-build/SimpleITK-build/CMakeTmp/link_flags/osx_dynamic_lookup/MODULE/SHARED/build/libnumber.dylib
15.7 ./SimpleITK-build/SimpleITK-build/CMakeTmp/link_flags/gnu_ld_ignore/MODULE/SHARED/build/libnumber.dylib
15.7 ./SimpleITK-build/CMakeTmp/link_flags/osx_dynamic_lookup/MODULE/SHARED/build/counter.so
15.7 ./SimpleITK-build/CMakeTmp/link_flags/osx_dynamic_lookup/MODULE/SHARED/build/libnumber.dylib
15.7 ./SimpleITK-build/CMakeTmp/link_flags/gnu_ld_ignore/MODULE/SHARED/build/libnumber.dylib
```

jcfr commented:
Considering that the `.dmg` package built on `computron` (`Slicer-5.9.0-2025-11-09`, `r34042`) is working as expected, we will revisit this later.
